### PR TITLE
embassy-net: add async wait_config_up

### PR DESCRIPTION
**Pull request by JuliDi**
This PR adds a function `wait_config_up` to the embassy_net `Stack`.
It will check whether `is_config_up` is true and if not register a waker that is woken in `apply_static_config` (e.g. by DHCP).

I added a not-working example to the docstring. Not sure whether we can make this run somehow.

### CODE REVIEW

File: embassy-net/src/lib.rs [hunk 1]
```diff
@@ -363,6 +366,56 @@ impl<D: Driver + 'static> Stack<D> {
         v4_up || v6_up
     }
 
+    /// Wait for the network stack to obtain a valid IP configuration.
+    /// Returns instantly if [`Stack::is_config_up`] returns `true`.
+    ///
+    /// ## Watch out:
```
Comment (MabezDev) [inline embassy-net/src/lib.rs:372]: I think this comment can be removed, or perhaps simplified to something like "Ensure `Stack::run` has been called" and defer explaining why to  `run`'s documentation.

**Comment by JuliDi** [comment]
Updated examples are tested on STM32H743

File: embassy-net/src/lib.rs [hunk 2]
```diff
@@ -363,6 +366,56 @@ impl<D: Driver + 'static> Stack<D> {
...
+    /// The Future is polled only when the [`Stack`] is running,
+    /// e.g. call `spawner.spawn(net_task(stack))`.
+    ///
+    /// `await`ing before will never yield!
```
Comment (MabezDev) [inline embassy-net/src/lib.rs:376]: There is a chance it will never yield anyway, for example in the case of DHCP. Maybe we should reword this to say as much, and that the end user should handle the timeout?

**Review by MabezDev — COMMENTED** [review]
Just a small drive-by review, feel free to ignore my comments. 

Very happy to see this added though, always bothered me when testing esp-wifi that the blocking versions resolved DHCP quicker :D.

File: embassy-net/src/lib.rs (hunk 2 above)
Comment (JuliDi) [reply embassy-net/src/lib.rs:376]: That's right. I guess if DHCP does not assign an IP, the user will have to handle it anyway. I'll think of something and maybe add an example.

Comment (JuliDi) [reply embassy-net/src/lib.rs]: I've rephrased the docstring. Do you think it is better now?

File: embassy-net/src/lib.rs [hunk 3]
```diff
@@ -363,6 +366,55 @@ impl<D: Driver + 'static> Stack<D> {
         v4_up || v6_up
     }
 
+    /// Wait for the network stack to obtain a valid IP configuration.
+    ///
+    /// ## Notes:
+    /// - Ensure [`Stack::run`] has been called before using this function.
+    ///
+    /// - This function may never yield (e.g. if no configuration is obtained through DHCP).
```
Comment (Dirbaio) [inline embassy-net/src/lib.rs:374]:
```suggestion
    /// - This function may never return (e.g. if no configuration is obtained through DHCP).
```

"yield" in async means the future suspending on an .await, not returning.

Comment (JuliDi) [reply embassy-net/src/lib.rs:374]: good catch, thanks

**Review by Dirbaio — APPROVED** [review]
thanks!
